### ISSUE TITLE
Bleadperl v5.17.4-246-g3465e1f breaks FREW/SQL-Translator-0.11016.tar.gz

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#12485]
Migrated from [rt.perl.org#115242](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=115242) (status was 'resolved')

Searchable as RT115242$

**Comment by p5pRT** [comment]
### From @andk

git bisect


---
commit 3465e1f03c6c748e8f8a6bf8bfdfaf1fc58a4810
Author&#8203;: Karl Williamson \<public@&#8203;khwilliamson\.com>
Date&#8203;:   Thu Oct 11 12&#8203;:15&#8203;:53 2012 \-0600

&nbsp;    regcomp\.c&#8203;: Optimize EXACTFish nodes without folds to EXACT

sample fail report


---
http&#8203;://www\.cpantesters\.org/cpan/report/65c62116\-154e\-11e2\-98ce\-0fa7b4e8b0a6

perl \-V


---
Summary of my perl5 \(revision 5 version 17 subversion 5\) configuration&#8203;:
&nbsp;  Commit id&#8203;: 3465e1f03c6c748e8f8a6bf8bfdfaf1fc58a4810
&nbsp;  Platform&#8203;:
&nbsp;    osname=linux\, osvers=3\.2\.0\-4\-amd64\, archname=x86\_64\-linux\-ld
&nbsp;    uname='linux k83 3\.2\.0\-4\-amd64 \#1 smp debian 3\.2\.30\-1 x86\_64 gnulinux '
&nbsp;    config\_args='\-Dprefix=/home/src/perl/repoperls/installed\-perls/perl/v5\.17\.4\-246\-g3465e1f/127e \-Dmyhostname=k83 \-Dinstallusrbinperl=n \-Uversiononly \-Dusedevel \-des \-Ui\_db \-Uuseithreads \-Duselongdouble \-DDEBUGGING=\-g'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    useithreads=undef\, usemultiplicity=undef
&nbsp;    useperlio=define\, d\_sfio=undef\, uselargefiles=define\, usesocks=undef
&nbsp;    use64bitint=define\, use64bitall=define\, uselongdouble=define
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, ccflags ='\-fno\-strict\-aliasing \-pipe \-fstack\-protector \-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
&nbsp;    optimize='\-O2 \-g'\,
&nbsp;    cppflags='\-fno\-strict\-aliasing \-pipe \-fstack\-protector \-I/usr/local/include'
&nbsp;    ccversion=''\, gccversion='4\.7\.2'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=8\, ptrsize=8\, doublesize=8\, byteorder=12345678
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
&nbsp;    ivtype='long'\, ivsize=8\, nvtype='long double'\, nvsize=16\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=16\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'\, ldflags =' \-fstack\-protector \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /lib/x86\_64\-linux\-gnu /lib/\.\./lib /usr/lib/x86\_64\-linux\-gnu /usr/lib/\.\./lib /lib /usr/lib
&nbsp;    libs=\-lnsl \-ldb \-ldl \-lm \-lcrypt \-lutil \-lc
&nbsp;    perllibs=\-lnsl \-ldl \-lm \-lcrypt \-lutil \-lc
&nbsp;    libc=\, so=so\, useshrplib=false\, libperl=libperl\.a
&nbsp;    gnulibc\_version='2\.13'
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-Wl\,\-E'
&nbsp;    cccdlflags='\-fPIC'\, lddlflags='\-shared \-O2 \-g \-L/usr/local/lib \-fstack\-protector'


Characteristics of this binary \(from libperl\)&#8203;: 
&nbsp;  Compile\-time options&#8203;: HAS\_TIMES PERLIO\_LAYERS PERL\_DONT\_CREATE\_GVSV
&nbsp;                        PERL\_MALLOC\_WRAP PERL\_PRESERVE\_IVUV PERL\_USE\_DEVEL
&nbsp;                        USE\_64\_BIT\_ALL USE\_64\_BIT\_INT USE\_LARGE\_FILES
&nbsp;                        USE\_LOCALE USE\_LOCALE\_COLLATE USE\_LOCALE\_CTYPE
&nbsp;                        USE\_LOCALE\_NUMERIC USE\_LONG\_DOUBLE USE\_PERLIO
&nbsp;                        USE\_PERL\_ATOF
&nbsp;  Built under linux
&nbsp;  Compiled at Oct 12 2012 05&#8203;:08&#8203;:39
&nbsp;  @&#8203;INC&#8203;:
&nbsp;    /home/src/perl/repoperls/installed\-perls/perl/v5\.17\.4\-246\-g3465e1f/127e/lib/site\_perl/5\.17\.5/x86\_64\-linux\-ld
&nbsp;    /home/src/perl/repoperls/installed\-perls/perl/v5\.17\.4\-246\-g3465e1f/127e/lib/site\_perl/5\.17\.5
&nbsp;    /home/src/perl/repoperls/installed\-perls/perl/v5\.17\.4\-246\-g3465e1f/127e/lib/5\.17\.5/x86\_64\-linux\-ld
&nbsp;    /home/src/perl/repoperls/installed\-perls/perl/v5\.17\.4\-246\-g3465e1f/127e/lib/5\.17\.5
&nbsp;    \.

\-\- 
andreas

**Comment by p5pRT** [comment]
### From @ilmari

\(Andreas J\. Koenig\) \(via RT\) \<perlbug\-followup@&#8203;perl\.org> writes&#8203;:

> git bisect
> \-\-\-\-\-\-\-\-\-\-
> commit 3465e1f03c6c748e8f8a6bf8bfdfaf1fc58a4810
> Author&#8203;: Karl Williamson \<public@&#8203;khwilliamson\.com>
> Date&#8203;:   Thu Oct 11 12&#8203;:15&#8203;:53 2012 \-0600
>
>     regcomp\.c&#8203;: Optimize EXACTFish nodes without folds to EXACT
>

This is caused by the following \(existing\, at least as far back as
5\.12\.4 \(the oldest I have handy\)\) regex optimizer bug&#8203;:

$ perl \-Mre=debug \-e '"/\\n\\n" =~ m\#\\A\(?&#8203;:^/$\)\#im'
Compiling REx "\\A\(?&#8203;:^/$\)"
Final program&#8203;:
&nbsp;   1&#8203;: SBOL \(2\)
&nbsp;   2&#8203;: MBOL \(3\)
&nbsp;   3&#8203;: EXACT \</> \(5\)
&nbsp;   5&#8203;: MEOL \(6\)
&nbsp;   6&#8203;: END \(0\)
anchored "/"$ at 0 \(checking anchored noscan\) anchored\(MBOL\)\(SBOL\) minlen 1 
Guessing start of match in sv for REx "\\A\(?&#8203;:^/$\)" against "/%n%n"
String too long\.\.\.
Match rejected by optimizer
Freeing REx&#8203;: "\\A\(?&#8203;:^/$\)"

The optimizer is ignoring the fact that MEOL can match more than one
character before the end of the string\.  The reason it only showed up
with the this commit is that the module was working around the bug by
using /i\, making the EXACT node an EXACTF\.

Thanks to khw for his help pinning it down on IRC\.

\-\- 
ilmari
"A disappointingly low fraction of the human race is\,
 at any given time\, on fire\." \- Stig Sandbeck Mathisen

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @ilmari

With /m in effect\, $ can match anywhere in the string\, and doesn't
say anything about the maximum length\.


---
 regcomp\.c     |    4 \+\+\+\-
 t/re/re\_tests |    1 \+
 2 files changed\, 4 insertions\(\+\)\, 1 deletion\(\-\)

<details><summary>Inline Patch</summary>

```diff
diff --git a/regcomp.c b/regcomp.c
index bf2be2d..0f134eb 100644
--- a/regcomp.c
+++ b/regcomp.c
@@ -5193,7 +5193,9 @@ S_setup_longest(pTHX_ RExC_state_t *pRExC_state, SV* sv_longest, SV** rx_utf8, S
                && (! meol || (RExC_flags & RXf_PMf_MULTILINE)))
           )
             /* See comments for join_exact for why REG_SEEN_EXACTF_SHARP_S */
-        || (RExC_seen & REG_SEEN_EXACTF_SHARP_S))
+        || (RExC_seen & REG_SEEN_EXACTF_SHARP_S)
+        /* MEOL can occur anywhere and says nothing about max length */
+        || meol)
     {
         return FALSE;
     }
diff --git a/t/re/re_tests b/t/re/re_tests
index a59a6ab..8021879 100644
--- a/t/re/re_tests
+++ b/t/re/re_tests
@@ -849,6 +849,7 @@ abb$	b\nca	n	-	-
 'abb\Z'm	b\nca	n	-	-
 'abb\z'm	b\nca	n	-	-
 'abb$'m	b\nca	n	-	-
+'\Aa$'m	a\n\n	y	$&	a
 (^|x)(c)	ca	y	$2	c
 a*abc?xyz+pqr{3}ab{2,}xy{4,5}pq{0,6}AB{0,}zz	x	n	-	-
 a(?{$a=2;$b=3;($b)=$a})b	yabz	y	$b	2
-- 
1.7.9.5
```
</details>

**Comment by p5pRT** [comment]
### From @demerphq

Ive looked into this patch a little  and it doesn't feel right\. 

While the assertion "MEOL says nothing about the maximum size of the 
string" is true it seems to go about implementing this by disabling 
longest required string detection and not by disabling the string length 
check for MEOL \(as apposed to SEOL\)\. 

BTW\, part of the trickiness here is figuring out where to start matching 
when we have foo$ which requires us to find the end of the string\, check 
if it is a newline then go back K characters \(3 in this case\) and check 
for "foo"\. The MEOL logic wants to work in the same way\, find the next 
newline or end of string then go back the required steps\. 

So if I am right and this logic disables this process then IMO it is not 
the right way to solve this bug\. Note however this is purely for textual 
analysis\. I haven't actually tried it to prove my case yet\.

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Ive looked into this patch a little  and it doesn't feel right\. 

While the assertion "MEOL says nothing about the maximum size of the 
string" is true it seems to go about implementing this by disabling 
longest required string detection and not by disabling the string length 
check for MEOL \(as apposed to SEOL\)\. 

BTW\, part of the trickiness here is figuring out where to start matching 
when we have foo$ which requires us to find the end of the string\, check 
if it is a newline then go back K characters \(3 in this case\) and check 
for "foo"\. The MEOL logic wants to work in the same way\, find the next 
newline or end of string then go back the required steps\. 

So if I am right and this logic disables this process then IMO it is not 
the right way to solve this bug\. Note however this is purely for textual 
analysis\. I haven't actually tried it to prove my case yet\.

**Comment by p5pRT** [comment]
### From @khwilliamson

Thanks for chasing this down\.  Yves Orton and I discussed your fix\, and
we both agree it goes too far\.  Your patch disables all substring length
checking\, both floating and anchored when a regex has a /m\.  It works\,
but it disables too much\, leading to slower regex matching unnecessarily\.

Instead\, I came up with a more modest fix\, just around the area where
the failure occurs\, to disable maximum string length checking under /m\.
 It does not affect the floating and anchored checks\.  Yves thought it
looked reasonable\.  It may be that we are wrong\, and more drastic
measures are called for\, but we don't have evidence of that at this time\.

I have committed my fix as d46b78c674c5ad23656932e71fea9369ee8ac4df
I used the test from your patch\, thank you very much\, as commit
fad493ef34530a95afba46f7b4f788dde6400695

\-\- 
Karl Williamson

**Comment by p5pRT** [comment]
@khwilliamson - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @ilmari

"Karl Williamson via RT" \<perlbug\-followup@&#8203;perl\.org> writes&#8203;:

> Thanks for chasing this down\.  Yves Orton and I discussed your fix\, and
> we both agree it goes too far\.  Your patch disables all substring length
> checking\, both floating and anchored when a regex has a /m\.  It works\,
> but it disables too much\, leading to slower regex matching unnecessarily\.

That was my main fear about the patch\, since i don't really know the
regex engine\.  Thanks for coming up with a more targeted fix\.

> Instead\, I came up with a more modest fix\, just around the area where
> the failure occurs\, to disable maximum string length checking under /m\.
>  It does not affect the floating and anchored checks\.  Yves thought it
> looked reasonable\.  It may be that we are wrong\, and more drastic
> measures are called for\, but we don't have evidence of that at this time\.
>
> I have committed my fix as d46b78c674c5ad23656932e71fea9369ee8ac4df
> I used the test from your patch\, thank you very much\, as commit
> fad493ef34530a95afba46f7b4f788dde6400695

\-\- 
"A disappointingly low fraction of the human race is\,
 at any given time\, on fire\." \- Stig Sandbeck Mathisen

**Comment by p5pRT** [comment]
### From @iabyn

Note that I mostly removed this fix with commit 6e88169\, since it was redundant now\.
